### PR TITLE
Add support for uploading proguard mapping files

### PR DESCRIPTION
**Pull request by tmtrademarked**
Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes:

- [ X ] Run `bundle exec rspec` from the subdirectory of each tool you modified. Alternatively, run `rake test_all` from the root directory.
- [ X ] Run `bundle exec rubocop -a` to ensure the code style is valid
- [ X ] Read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [ X ] We currently don't accept new actions, please publish a plugin instead, more information in [Plugins.md](https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Plugins.md)

Before submitting a pull request, we appreciate if you create an issue first to discuss the change :+1:

This PR adds functionality to support adding a proguard mapping file per apk. This will allow developers to upload their deobfuscation files at the same time as they upload their APKs, making life better for everyone.

This is an attempt to implement the functionality requested in https://github.com/fastlane/fastlane/issues/7183.

Note that I haven't tested this functionality yet - any comments on the best way to write unit tests for this and to replace my currently installed Supply with my locally built version would be much appreciated!

### CODE REVIEW

**Comment by joreilly** [comment]
Really looking forward to this being available!

**Comment by hjanuschka** [comment]
@tmtrademarked 

Maybe one of the below option's help you to run your modified supply.

# Option 1
get this script into your project folder, 
https://gist.github.com/hjanuschka/3040fc6f371f46acff72a5e67f649b27

This requires you to push changes first to github and then call `try_pr.sh` again, 
but its an easy way to every one trien out a PR.

```bash
cd /path/to/your/project/
# this only has to be done once.
curl -o try_pr.sh https://gist.githubusercontent.com/hjanuschka/3040fc6f371f46acff72a5e67f649b27/raw/9600a34188949061a96f77c82505efd99ad6f5e4/try_pr.sh
chmod a+x  try_pr.sh


# then call it with (which in this case should show the help screen of your supply):
./try_pr.sh tmtrademarked proguard_files supply --help
```


# Option 2
Totally local.:

```bash
cd /path/to/your/fastlane-clone/
# ensure your branch is checked out
export FASTLANE_LOCAL_DEV=1
bundle update # only once
bundle install #only once
bundle exec supply --help
```


> just to be sure, unsupported, unofficial, use at your own risk! - and review the shell script before you run it.

File: supply/lib/supply/options.rb
Comment (hjanuschka) [inline supply/lib/supply/options.rb]: this one raises an exception:

```
fastlane_core/lib/fastlane_core/ui/interface.rb:145:in `user_error!': short_option must be a String of length 1 (FastlaneCore::Interface::FastlaneError)
```

you should use something different than, `-mp`

**Comment by tmtrademarked** [comment]
Thanks, @hjanuschka - that worked like a charm! In order to minimize noise, I ended up using your "Option 2" approach. Having tested this with my build, I can now verify that the latest PR uploads mapping files to the play console correctly.

PTAL and let me know what you think of the approach.

**Comment by hjanuschka** [comment]
@tmtrademarked thx for feedback, glad you got it rolling!
lets wait for the core team, to review it.

**Comment by tmtrademarked** [comment]
Hey, all - just wanted to check in on the status here. Any thoughts about the feature in general, or the implementation in specific?

**Comment by asfalcone** [comment]
@tmtrademarked Thanks for the PR. The idea is great and the code is nice and simple. I am hoping to find some time to pull down the code and test it out before approving. 

I'd like to see a `default_value` for the mapping path. 

If there are other folks who can manually test this, that would be super helpful!

**Comment by tmtrademarked** [comment]
Thanks for the comment, @asfalcone ! Re: default value - the reason I didn't add one here was that I thought it would be somewhat complicated. It's pretty easy to look in the default path for app/release for a mapping.txt file, but if that file doesn't exist, it will cause the "file exists" validation to fail. Some developers may not be using proguard, for example.

So it seemed better to ask developers to explicitly opt in to the mapping file upload by specifying the path. Is there a better way to do the validation that could distinguish the default from a user provided option?

**Comment by tmtrademarked** [comment]
Quick ping on this one - any further refinements/alterations/etc people would like to see?

**Comment by asfalcone** [comment]
Sorry for the delay! I was off for the holidays! 🎄 ⭐️

File: supply/lib/supply/uploader.rb
Comment (asfalcone) [inline supply/lib/supply/uploader.rb]: Can we remove the `puts` here? 

If you think the output is valuable, provide a nicely formatted string like `UI.message "Uploading #{mapping_path} for #{version_code}."`

**Review by asfalcone — CHANGES_REQUESTED** [review]

File: supply/lib/supply/uploader.rb
Comment (tmtrademarked) [inline supply/lib/supply/uploader.rb]: Whoops - those were just debugging code. Thanks for the reminder! Removed.

**Review by asfalcone — APPROVED** [review]

**Comment by asfalcone** [comment]
Looks like we've got some failing tests. Probably just needs to be rebased on master @tmtrademarked

**Comment by tmtrademarked** [comment]
I thought I did the rebasing correctly, but it looks like I may have messed something up.

**Comment by tmtrademarked** [comment]
This looks like it may be tricky to recover from. Let me close this PR and recreate from a fresh fork, which will be simpler. (Any tips on how to do this sort of thing correctly in the future are appreciated, though!)

**Comment by asfalcone** [comment]
#7721 replaces this

**Pull request closed without merging**
